core/core-impl/src/main/java/io/fabric8/launcher/core/impl/identity/KeycloakPublicKeyProvider.java

```diff
@@ -0,0 +1,110 @@
+
+    private final KeycloakParameters keycloakParameters;
+
+    private final TokenIdentity identity;
+
+    private final HttpClient httpClient;
+
+    @Inject
+    public KeycloakPublicKeyProvider(final KeycloakParameters keycloakParameters, TokenIdentity identity, final HttpClient httpClient) {
+        this.keycloakParameters = Objects.requireNonNull(keycloakParameters, "keycloakParameters must be specified");
+        this.identity = Objects.requireNonNull(identity, "Token identity must be specified");
+        this.httpClient = Objects.requireNonNull(httpClient, "httpClient must be specified");
+    }
+
+    @Override
+    public Optional<RSAPublicKey> getKey(String keyId) {
+        final String url = this.keycloakParameters.buildKeysUrl();
+        final Request request = securedRequest(identity)
+                .url(url)
+                .build();
+        try {
+            final Map<String, RSAPublicKey> publicKeys = httpClient.executeAndMap(request, KeycloakPublicKeyProvider::findKeys);
+            final RSAPublicKey publicKey = Objects.requireNonNull(publicKeys).get(keyId);
+
+            return Optional.ofNullable(publicKey);
+        } catch (final Exception e) {
+            logger.log(Level.SEVERE, "Error while fetching keys from keycloak for kid: " + keyId, e);
+            return Optional.empty();
+        }
+    }
+
+    private static Map<String, RSAPublicKey> findKeys(Response r) {
+        try (final ResponseBody body = r.body()) {
+            final String content = getContent(body);
+            final JsonNode node = JsonUtils.readTree(content);
+            if (!r.isSuccessful()) {
+                throw new IllegalStateException(extractFieldFromNodeOrDefaultTo(node, "errorMessage", ""));
+            }
+            return findAllPublicKeys(node);
+        } catch (final IOException e) {
+            throw new IllegalStateException("Error while fetching token from keycloak", e);
+        }
+    }
+
+    private static Map<String, RSAPublicKey> findAllPublicKeys(JsonNode node) {
+        final Map<String, RSAPublicKey> publicKeys = new HashMap<>();
+        if (!node.hasNonNull("keys")) {
+            logger.severe(String.format("Expected 'keys' to be present in the response:\n %s", node.asText()));
+            return Collections.emptyMap();
+        }
+        node.get("keys")
+                .iterator()
+                .forEachRemaining(keyNode -> {
+                                      final RSAPublicKey publicKey = transformToRsa(keyNode);
+                                      publicKeys.put(extractFieldFromNodeOrDefaultTo(keyNode, "kid", "kid"), publicKey);
+                                  }
+                );
+        return publicKeys;
+    }
+
+    private static RSAPublicKey transformToRsa(JsonNode keyNode) {
+        final String modulus = extractFieldFromNodeOrDefaultTo(keyNode, "n", null);
+        final String exponent = extractFieldFromNodeOrDefaultTo(keyNode, "e", null);
```

Comment:
these two values are only used if we didn't fail with the exception - maybe we should move it down to the place where they are being used so we don't calculate them unnecessarily upon failure